flytekit/annotated/context_manager.py

```diff
@@ -0,0 +1,101 @@
+from contextlib import contextmanager
+from typing import List, Optional
+
+from flytekit.configuration import sdk as _sdk_config, platform as _platform_config
+from flytekit.interfaces.data.s3 import s3proxy as _s3proxy
+from flytekit.interfaces.data.gcs import gcs_proxy as _gcs_proxy
+from flytekit.interfaces.data.local import local_file_proxy as _local_file_proxy
+from flytekit.interfaces.data.http import http_data_proxy as _http_data_proxy
+from flytekit.common.exceptions import user as _user_exception
+from flytekit.common import utils as _common_utils, constants as _constants
+from flytekit.common.nodes import SdkNode
+from flytekit.interfaces.data import common as _common_data
+from flytekit.clients import friendly as flyte_client
+
+
+class CompilationState(object):
+    def __init__(self):
+        self.nodes: List[SdkNode] = []
+        self.mode = 1  # TODO: Turn into enum in the future, or remove if only one mode.
+
+
+class FlyteContext(object):
+    OBJS = []
+
+    def __init__(self, parent=None,
+                 local_file_access: _local_file_proxy.LocalFileProxy = None,
+                 remote_data_proxy: _common_data.DataProxy = None,
+                 compilation_state: CompilationState = None,
+                 flyte_client: flyte_client.SynchronousFlyteClient = None,
+                 ):
+        # TODO: Should we have this auto-parenting feature?
+        if parent is None and len(FlyteContext.OBJS) > 0:
+            parent = FlyteContext.OBJS[-1]
+
+        self._parent: FlyteContext = parent
+        self._local_file_access = local_file_access
+        self._remote_data_proxy = remote_data_proxy
+        self._compilation_state = compilation_state
+        self._flyte_client = flyte_client
+
+    def __enter__(self):
+        FlyteContext.OBJS.append(self)
```

Comment:
Semantically won't matter but should this __enter__ forcefully set `self._parent = FlyteContext.OBJS[-1]`

Reply:
you decide.

Reply:
the whole array keeping track of everything is annoying actually, i'd rather not have it.  (but I think it's preferable to using thread locals).  the only place it's really used (read as, "the only place where context isn't explicitly passed as as an arg"), is when a task is called, since we don't inject a context arg anymore (which is a good thing imo btw).